### PR TITLE
Replace CNMeM (and some test cleanup)

### PR DESCRIPTION
**Pull request by harrism**
This PR:

 - [x] Replaces usage of deprecated (as of RMM 0.15) `cnmem_memory_resource` with `pool_memory_resource` and uses it as the default.
 - [x] Adds a tests/utilities/base_fixture.hpp modeled after cudf
 - [x] Adds a standard gtest main function that accepts command line choice of the RMM resource to use
 - [x] Cleans up the messy includes in all gtests
 - [x] Replaces all use of thrust::device_vector in gtests with rmm::device_vector
 - [x] includes a new third party header, cxxopts.hpp, which provides command line parsing capability to gtests (to allow choosing the RMM resource from the command line).
- [x] Update all tests to use new gtest main function macro and clean up includes and device_vector usage.
- [x] Add base_fixture.hpp and cxxopts.hpp

### CODE REVIEW

File: cpp/tests/utilities/cxxopts.hpp [hunk 1]
```diff
@@ -0,0 +1,1497 @@
...
+  std::vector<std::string>::iterator m_next_positional;
+  std::unordered_set<std::string> m_positional_set;
+  std::unordered_map<std::shared_ptr<OptionDetails>, OptionValue> m_results;
+
+  bool m_allow_unrecognised;
+
+  std::vector<KeyValue> m_sequential;
+};
+
+struct Option {
```
Comment (aschaffer) [inline cpp/tests/utilities/cxxopts.hpp:835]: Type naming (constructors, destructors, template params, etc.) convention we use is lowercase `<name>_t`. Also, please use more descriptive names.

File: cpp/tests/utilities/cxxopts.hpp [hunk 2]
```diff
@@ -0,0 +1,1497 @@
...
+  {
+  }
+
+  std::string opts_;
+  std::string desc_;
+  std::shared_ptr<const Value> value_;
+  std::string arg_help_;
+};
+
+class Options {
```
Comment (aschaffer) [inline cpp/tests/utilities/cxxopts.hpp:850]: Type naming (constructors, destructors, template params, etc.) convention we use is lowercase `<name>_t`. Also, please use more descriptive names. Please provide a name that clearly differentiates between an `Option` abstraction and (I assume) a collection, `Options`. Just One "s" is not informative enough.

File: cpp/tests/utilities/cxxopts.hpp [hunk 3]
```diff
@@ -0,0 +1,1497 @@
...
+  std::shared_ptr<OptionMap> m_options;
+  std::vector<std::string> m_positional;
+  std::vector<std::string>::iterator m_next_positional;
+  std::unordered_set<std::string> m_positional_set;
+
+  // mapping from groups to help options
+  std::map<std::string, HelpGroupDetails> m_help;
+};
+
+class OptionAdder {
```
Comment (aschaffer) [inline cpp/tests/utilities/cxxopts.hpp:949]: Type naming (constructors, destructors, template params, etc.) convention we use is lowercase `<name>_t`.

File: cpp/tests/utilities/cxxopts.hpp [hunk 4]
```diff
@@ -0,0 +1,1497 @@
...
+
+namespace {
+constexpr int OPTION_LONGEST  = 30;
+constexpr int OPTION_DESC_GAP = 2;
+
+std::basic_regex<char> option_matcher("--([[:alnum:]][-_[:alnum:]]+)(=(.*))?|-([[:alnum:]]+)");
+
+std::basic_regex<char> option_specifier("(([[:alnum:]]),)?[ ]*([[:alnum:]][-_[:alnum:]]*)?");
+
+String format_option(const HelpOptionDetails& o)
```
Comment (aschaffer) [inline cpp/tests/utilities/cxxopts.hpp:973]: Type naming (constructors, destructors, template params, etc.) convention we use is lowercase `<name>_t`.

File: cpp/tests/utilities/cxxopts.hpp [hunk 5]
```diff
@@ -0,0 +1,1497 @@
...
+      result += " [=" + arg + "(=" + toLocalString(o.implicit_value) + ")]";
+    } else {
+      result += " " + arg;
+    }
+  }
+
+  return result;
+}
+
+String format_description(const HelpOptionDetails& o, size_t start, size_t width)
```
Comment (aschaffer) [inline cpp/tests/utilities/cxxopts.hpp:1001]: Type naming (constructors, destructors, template params, etc.) convention we use is lowercase `<name>_t`.

File: cpp/tests/utilities/cxxopts.hpp [hunk 6]
```diff
@@ -0,0 +1,1497 @@
...
+          // parse the next argument
+          checked_parse_arg(argc, argv, current, opt, name);
+        }
+      }
+    }
+
+    ++current;
+  }
+
+  for (auto& opt : *m_options) {
```
Comment (aschaffer) [inline cpp/tests/utilities/cxxopts.hpp:1317]: Consider using r-value reference in range loops: `for (auto&& opt: ...)` is more generic and arguably more efficient.

File: cpp/tests/utilities/cxxopts.hpp [hunk 7]
```diff
@@ -0,0 +1,1497 @@
...
+// which results in the correct lengths being computed for strings when they
+// are formatted for the help output
+// it is necessary to make sure that <unicode/unistr.h> can be found by the
+// compiler, and that icu-uc is linked in to the binary.
+
+#ifdef CXXOPTS_USE_UNICODE
+#include <unicode/unistr.h>
+
+namespace cxxopts {
+typedef icu::UnicodeString String;
```
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/tests/utilities/cxxopts.hpp:61]: cxxopts is a vendored open-source single-header library that we use in libcudf and RMM as well. We aren't going to change it.

File: cpp/tests/utilities/cxxopts.hpp [hunk 8]
```diff
@@ -0,0 +1,1497 @@
...
+#ifdef _WIN32
+const std::string LQUOTE("\'");
+const std::string RQUOTE("\'");
+#else
+const std::string LQUOTE("‘");
+const std::string RQUOTE("’");
+#endif
+}  // namespace
+
+class Value : public std::enable_shared_from_this<Value> {
```
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/tests/utilities/cxxopts.hpp:180]: Again, third party library.

**Comment by harrism** [comment]
@aschaffer I'm sorry you spent all that time writing 30+ comments.

I should have noted in the description that this adds a third party single-header library, [cxxopts](https://github.com/jarro2783/cxxopts) for command line argument parsing. This was so we can use the exact same gtest main code as libcudf. 

Also I realized I put in code that uses `binning_memory_resource`, which is in a PR under review in RMM 0.15. So I might comment that out for now. (that's why CI is failing).

**Comment by harrism** [comment]
CI will pass once rapidsai/rmm#457 is merged.

**Comment by BradReesWork** [comment]
rerun tests

**Comment by BradReesWork** [comment]
rerun tests

**Comment by BradReesWork** [comment]
rerun tests

**Comment by BradReesWork** [comment]
rerun tests

**Comment by BradReesWork** [comment]
rerun tests

**Pull request merged**